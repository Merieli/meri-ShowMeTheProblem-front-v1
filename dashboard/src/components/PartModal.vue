<template>
    <teleport to="body">
        <div
            class="overlay fixed top-0 left-0 z-1 flex items-center justify-center w-full h-full bg-black bg-opacity-50"
            v-show="open"
            @click="closeModal"
        >
            <transition name="moda-animation-inner">
                <div class="fixed mx-10" @click.stop>
                    <div
                        class="modal w-96 z-10 flex flex-col overflow-hidden bg-white rounded-lg .animate__animated animate__faster animate__fadeInDown"
                    >
                        <div class="modal__header mx-8 my-8">
                            <legend v-if="typeModal == 'create'" class="font-black text-3xl">Crie uma conta</legend>
                            <legend v-else class="font-black text-3xl">Entre na sua conta</legend>
                            <button @click="closeModal" class="modal__button">
                                <svg xmlns="http://www.w3.org/2000/svg" height="24" width="24">
                                    <path
                                        d="M6.4 19 5 17.6l5.6-5.6L5 6.4 6.4 5l5.6 5.6L17.6 5 19 6.4 13.4 12l5.6 5.6-1.4 1.4-5.6-5.6Z"
                                    />
                                </svg>
                            </button>
                        </div>
                        <div class="modal__content">
                            <form
                                v-if="typeModal == 'create'"
                                class="modal__form mx-12 my-10 mt-0"
                                @submit.prevent="createUser"
                            >
                                <fieldset class="flex flex-col">
                                    <label for="name" class="font-bold text-lg" id="create-name"> Nome </label>
                                    <input
                                        class="bg-slate-50 p-3 text-lg border-2 border-transparent rounded"
                                        :class="{
                                            'border-brand-danger': !!state.name.errorMessage,
                                        }"
                                        type="text"
                                        name="name"
                                        placeholder="Seu Nome Aqui"
                                        v-model="state.name.value"
                                        autocomplete="on"
                                        required
                                        data-form-name
                                    />
                                    <span v-if="!!state.name.errorMessage" class="block font-medium text-brand-danger">
                                        {{ state.name.errorMessage }}
                                    </span>
                                    <label for="email" class="font-bold text-lg mt-6">E-mail</label>
                                    <input
                                        class="bg-slate-50 p-3 text-lg border-2 border-transparent rounded"
                                        :class="{
                                            'border-brand-danger': !!state.email.errorMessage,
                                        }"
                                        type="email"
                                        name="email"
                                        placeholder="email@exemplo.com"
                                        v-model="state.email.value"
                                        autocomplete="on"
                                        required
                                    />
                                    <span v-if="!!state.email.errorMessage" class="block font-medium text-brand-danger">
                                        {{ state.email.errorMessage }}
                                    </span>
                                    <label for="password" class="font-bold text-lg mt-6">Senha</label>
                                    <input
                                        class="bg-slate-50 p-3 text-lg border-2 border-transparent rounded"
                                        :class="{
                                            'border-brand-danger': !!state.password.errorMessage,
                                        }"
                                        type="password"
                                        name="password"
                                        placeholder="******"
                                        v-model="state.password.value"
                                        required
                                    />
                                    <span
                                        v-if="!!state.password.errorMessage"
                                        class="block font-medium text-brand-danger"
                                    >
                                        {{ state.password.errorMessage }}
                                    </span>
                                </fieldset>

                                <BaseButton
                                    data-button="createAccount"
                                    color="dark"
                                    type-button="submit"
                                    :disabled="isLoading"
                                    class="mt-6"
                                    :class="{
                                        'opacity-50': isLoading,
                                    }"
                                >
                                    Criar conta
                                </BaseButton>
                            </form>
                            <form
                                v-else
                                class="modal__form mx-12 my-10 mt-0"
                                @submit.prevent="loginUser"
                                data-modal-form
                            >
                                <fieldset class="flex flex-col">
                                    <label for="email" class="font-bold text-lg">E-mail</label>
                                    <input
                                        class="bg-slate-50 p-3 text-lg border-2 border-transparent rounded"
                                        :class="{
                                            'border-brand-danger': !!state.email.errorMessage,
                                        }"
                                        type="email"
                                        name="email"
                                        placeholder="email@exemplo.com"
                                        v-model="state.email.value"
                                        autocomplete="on"
                                        required
                                    />
                                    <span v-if="!!state.email.errorMessage" class="block font-medium text-brand-danger">
                                        {{ state.email.errorMessage }}
                                    </span>
                                    <label for="password" class="font-bold text-lg mt-6">Senha</label>
                                    <input
                                        class="bg-slate-50 p-3 text-lg border-2 border-transparent rounded"
                                        :class="{
                                            'border-brand-danger': !!state.password.errorMessage,
                                        }"
                                        type="password"
                                        name="password"
                                        placeholder="******"
                                        v-model="state.password.value"
                                        required
                                    />
                                    <span
                                        v-if="!!state.password.errorMessage"
                                        class="block font-medium text-brand-danger"
                                    >
                                        {{ state.password.errorMessage }}
                                    </span>
                                </fieldset>

                                <BaseButton color="dark" type-button="submit" :disabled="isLoading" class="mt-6">
                                    Entrar
                                </BaseButton>
                            </form>
                        </div>
                    </div>
                </div>
            </transition>
        </div>
    </teleport>
</template>

<script lang="ts">
/** @version 1.0.0
 */
import { IAccount } from '../interfaces'
import { useStore } from '../store'
import { Actions } from '../store/type-actions'
import { validateEmptyAndLenght6, validateEmptyAndEmail } from '../utils/validators'
import BaseButton from './BaseButton/index.vue'
import { useField } from 'vee-validate'
import { defineComponent, computed, reactive, ref } from 'vue'
import * as yup from 'yup'

export default defineComponent({
    name: 'PartModal',
    emits: ['close', 'open'],
    components: {
        BaseButton,
    },
    props: {
        /**
         * Define o tipo do modal a ser usado
         * @values create, login
         */
        typeModal: {
            type: String,
            default: 'create',
        },
        /**
         * Define se o modal está sendo exibido ou não.
         * @values true, false
         */
        open: {
            type: Boolean,
            default: false,
        },
    },
    setup(props, { emit }) {
        const store = useStore()
        const isActive = ref(true)

        const { value: nameValue, errorMessage: nameErrorMessage } = useField<string>(
            'name',
            yup.string().required().min(6)
        )

        const { errorMessage: emailErrorMessage, value: emailValue } = useField<string>('email', validateEmptyAndEmail)

        const { errorMessage: passwordErrorMessage, value: passwordValue } = useField<string>(
            'password',
            validateEmptyAndLenght6
        )

        const state: IAccount = reactive({
            name: {
                value: nameValue,
                errorMessage: nameErrorMessage,
            },
            email: {
                value: emailValue,
                errorMessage: emailErrorMessage,
            },
            password: {
                value: passwordValue,
                errorMessage: passwordErrorMessage,
            },
        })

        /** Acionado quando o modal tiver o valor true.
         * @event open
         * @type {Event}
         */
        const showModal = (active: boolean) => {
            isActive.value = active
            emit('open', active)
        }

        /** Acionado para fechar o modal.
         * @event open
         * @type {Event}
         */
        const closeModal = () => {
            emit('close')
        }

        /** Limpa os campos do modal
         * @event success
         */
        const cleanAll = () => {
            state.name.value = ''
            // state.name.errorMessage = ''

            state.email.value = ''
            // state.email.errorMessage = ''

            state.password.value = ''
            // state.password.errorMessage =
        }

        const createUser = () => {
            store
                .dispatch(Actions.REGISTER_USER, {
                    name: state.name.value,
                    email: state.email.value,
                    password: state.password.value,
                })
                .then(() => {
                    cleanAll()
                    closeModal()
                })
        }

        const loginUser = () => {
            store
                .dispatch(Actions.LOGIN_USER, {
                    email: state.email.value,
                    password: state.password.value,
                })
                .then(() => {
                    cleanAll()
                    closeModal()
                })
        }

        return {
            state,
            isActive,
            isLoading: computed(() => store.state.isLoading),
            createUser,
            loginUser,
            showModal,
            closeModal,
        }
    },
})
</script>

<style lang="scss" scoped>
.modal {
    z-index: 2;

    &__header {
        display: flex;
        justify-content: space-between;
        align-items: center;
    }

    &__button {
        transition: all 200ms ease;
        padding: 0.4rem;

        &:hover {
            background-color: gray;
            transform: rotate(-180deg);

            svg {
                fill: white;
            }
        }
    }
}
.overlay {
    z-index: 0;
}
</style>

<docs lang="md">
Modal que pode ser utilizado para login ou criação de conta de usuário.

## Exemplo

Modal de Login:

```vue
<PartModal typeModal="login" @open="true" />
```
</docs>
