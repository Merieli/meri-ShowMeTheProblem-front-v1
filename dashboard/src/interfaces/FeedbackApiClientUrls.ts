export interface FeedbackApiClientUrls {
    create: string
    show: string
    getFilters: string
}
