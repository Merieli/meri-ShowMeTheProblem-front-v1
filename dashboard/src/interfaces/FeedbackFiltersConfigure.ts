import { IFeedbackFilters, TFeedback } from './'

export interface FeedbackFiltersConfigure {
    data: IFeedbackFilters
    typeActive: TFeedback
}
