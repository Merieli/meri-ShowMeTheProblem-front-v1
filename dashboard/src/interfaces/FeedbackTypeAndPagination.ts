import { TFeedback } from './'

export interface FeedbackTypeAndPagination {
    type: TFeedback
    limit: number
    offset: number
}
