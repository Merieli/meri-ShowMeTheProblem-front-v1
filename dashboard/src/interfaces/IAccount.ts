export interface IAccount {
    name: {
        value?: string
        errorMessage?: string
    }
    email: {
        value?: string
        errorMessage?: string
    }
    password: {
        value?: string
        errorMessage?: string
    }
}
