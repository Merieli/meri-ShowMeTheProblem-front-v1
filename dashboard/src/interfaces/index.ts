import { IAccount } from './IAccount'

export * from './'
