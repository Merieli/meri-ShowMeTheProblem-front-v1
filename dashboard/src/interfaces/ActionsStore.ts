import { FeedbackTypeAndPagination, IUser, StateStoreShape, TFeedback } from '.';
import { Actions } from '@/store/type-actions';
import { ActionContext, ActionTree } from 'vuex';

type Context = ActionContext<StateStoreShape, StateStoreShape>;

export interface ActionsStore extends ActionTree<StateStoreShape, StateStoreShape> {
    /**
     * @name REGISTER_USER
     * @descripton efetua o registro do usuário e o adiciona a store
     * @param {IUser} user usuário com name, email e password
     */
    [Actions.REGISTER_USER]: ({ commit }: Context, user: IUser) => Promise<void>;

    /**
     * @name LOGIN_USER
     * @description Efetua login do usuário, salva o token na store,
     * redireciona para a página de 'feedbacks' e aciona a action
     * para pegar os dados do usuário logado.
     * @param {IUser} user usuário com name, email e password
     */
    [Actions.LOGIN_USER]: (context: Context, user: IUser) => Promise<void>;

    /**
     * @name GET_USER
     * @descripton Captura os dados do usuário e efetua o login se já tiver um
     * token de usuário salvo, armazenando os dados na store
     */
    [Actions.GET_USER]: (context: Context) => Promise<void>;

    /**
     * @name LOGGOUT_USER
     * @descripton faz o usuário ser deslogado
     */
    [Actions.LOGGOUT_USER]: (context: Context) => Promise<void>;

    /**
     * @name GENERATE_NEW_KEY
     * @descripton gera uma nova chave de api para o usuário e a salva na store
     */
    [Actions.GENERATE_NEW_KEY]: (context: Context) => Promise<void>;

    [Actions.HANDLE_COPY_TEXT]: (context: Context, text: string) => Promise<void>;

    /**
     * @name GET_SUMMARY_FEEDBACK
     * @descripton pega a quantidade de cada tipo de feedback e salva na store
     * configurando para exibir os filtros da listagem
     */
    [Actions.GET_SUMMARY_FEEDBACK]: (context: Context) => Promise<void>;

    /**
     * @name CHANGE_ACTIVE_FEEDBACK
     * @descripton altera o tipo de feedback ativo
     * @param {TFeedback} type tipo do feedback retornado
     */
    [Actions.CHANGE_ACTIVE_FEEDBACK]: (context: Context, type: TFeedback) => Promise<void>;

    /**
     * @name GET_ALL_FEEDBACKS
     * @descripton
     * @param {Object} paramsRequest
     */
    [Actions.GET_ALL_FEEDBACKS]: (
        context: Context,
        feedbackTypeAndPagination: FeedbackTypeAndPagination
    ) => Promise<void>;
}
