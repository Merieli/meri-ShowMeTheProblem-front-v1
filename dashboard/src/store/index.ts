import useNotifier from '../hooks/notifier'
import router from '../router/index'
import { Actions } from './type-actions'
import { Mutations } from './type-mutations'
import callApiClient from '@/http-client/index'
import { TypeOfNotification } from '@/interfaces/INotification'
import { INotification } from '@/interfaces/INotification'
import IUser from '@/interfaces/IUser'
import IUserLogged from '@/interfaces/IUserLogged'
import { createStore } from 'vuex'

export interface EstadoStore {
    users: IUser[]
    notifications: INotification[]
    isLogged: boolean
    userLogged: IUserLogged
}

export const store = createStore<EstadoStore>({
    state: {
        users: [],
        notifications: [],
        isLogged: false,
        userLogged: {
            name: '',
            token: '',
            apiKey: '',
        },
    },
    getters: {},
    mutations: {
        [Mutations.NOTIFY](state, newNotification: INotification) {
            newNotification.id = new Date().getTime()
            state.notifications.push(newNotification)
            setTimeout(() => {
                state.notifications = state.notifications.filter((notif) => notif.id != newNotification.id)
            }, 5000)
        },
        [Mutations.ADD_USER](state, user: IUser) {
            state.users.push(user)
        },
        [Mutations.LOGIN_USER](state, token) {
            state.userLogged.token = token
            state.isLogged = true
        },
        [Mutations.USER_LOGGED](state, data) {
            state.userLogged.name = data.name
            state.userLogged.apiKey = data.apiKey
        },
    },
    actions: {
        /**
         * @name REGISTER_USER
         * @descripton efetua o registro do usuário e o adiciona a store
         * @param {IUser} user usuário com name, email e password
         */
        async [Actions.REGISTER_USER]({ commit }, user: IUser) {
            try {
                const response = await callApiClient.user.register(user.name, user.email, user.password)
                commit(Mutations.ADD_USER, response)
            } catch (e) {
                throw new Error(`Não foi possível efetuar cadastro. Confira os dados informados`)
            }
        },

        /**
         * @name LOGIN_USER
         * @description Efetua login do usuário, salva o token na store,
         * redireciona para a página de 'feedbacks' e aciona a action
         * para pegar os dados do usuário logado.
         * @param {IUser} user usuário com name, email e password
         */
        async [Actions.LOGIN_USER]({ commit, dispatch, state }, user: IUser) {
            try {
                const response = await callApiClient.user.login(user.email, user.password)
                commit(Mutations.LOGIN_USER, response.token)
                router.push('/feedbacks')
                dispatch(Actions.GET_USER, state.userLogged.token)
            } catch (e) {
                const { notify } = useNotifier()
                notify(TypeOfNotification.FALHA, 'Não foi possível efetuar login', 'Confira os dados informados.')
                throw new Error(`Não foi possível efetuar login. Confira os dados informados`)
            }
        },

        /**
         * @name GET_USER
         * @descripton Captura os dados do usuários e salva na store
         * @param token chave de acesso do usuário
         */
        async [Actions.GET_USER]({ commit }, token) {
            try {
                const response = await callApiClient.user.show(token)
                commit(Mutations.USER_LOGGED, response)
            } catch (e) {
                throw new Error(`Não foi possível capturar os dados do usuário.`)
            }
        },
    },
    modules: {},
})
