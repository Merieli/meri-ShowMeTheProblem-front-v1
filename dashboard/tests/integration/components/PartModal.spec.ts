import PartModal from '@/components/PartModal.vue'
import { StateStoreShape } from '@/interfaces'
import { key } from '@/store'
import { mount } from '@vue/test-utils'
import { createStore, Store } from 'vuex'

describe('PartModal', () => {
    jest.mock('@/http/index')
    const mockLoading = () => mockState.isLoading
    type TStoreState = Pick<StateStoreShape, 'users' | 'isLoading'>

    const mockState: TStoreState = {
        users: [
            {
                name: 'SUCESSO',
                email: 'Teste',
                password: 'Uma notificação exibida',
            },
        ],
        isLoading: false,
    }
    let mockVuexStore: Store<TStoreState>

    beforeEach(() => {
        mockVuexStore = createStore<TStoreState>({
            state: mockState,
            getters: { notifications: mockLoading },
            actions: {},
        })
    })

    function factory(storeMock: Store<TStoreState>, type: string) {
        return mount(PartModal, {
            global: {
                plugins: [[storeMock, key]],
                stubs: {
                    teleport: true,
                    transition: true,
                },
            },
            props: {
                open: true,
                typeModal: type,
            },
        })
    }

    describe('Integração/Componente', () => {
        describe('Renderização:', () => {
            test('Dado o html quando renderizado então deve ter os mesmos dados do snapshot gravado', () => {
                const wrapper = factory(mockVuexStore, 'create')
                expect(wrapper.html()).toMatchSnapshot()
                console.log('>>>>', wrapper.html())
            })
        })
        describe('Comportamento:', () => {
            return
        })
        describe('Navegação:', () => {
            return
        })
    })
})
