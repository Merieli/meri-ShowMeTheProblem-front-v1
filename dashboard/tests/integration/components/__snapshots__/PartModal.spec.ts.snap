// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PartModal Integração/Componente Renderização: Dado o html quando renderizado então deve ter os mesmos dados do snapshot gravado 1`] = `
<!--teleport start-->
<!--teleport end-->
`;
